5-WebApp-AuthZ-Roles-Groups/README.md

```diff
@@ -0,0 +1,252 @@
+        }
+    }
+  ...
+}
+```
+> An Identity Office Hours session covered Azure AD App roles and security groups, featuring this scenario and how to handle the overage claim. Watch the video [Using Security Groups and Application Roles in your apps](https://www.youtube.com/watch?v=V8VUPixLSiM)
+
+ > You can use the `BulkCreateGroups.ps1` provided in the [App Creation Scripts](./AppCreationScripts/) folder to help test overage scenarios.
+
+##### Order of processing the overage claim
+
+1. Check for the claim `_claim_names` with one of the values being `groups`. This indicates overage.
+
+1. If found, make a call to the endpoint specified in `_claim_sources` to fetch user’s groups.
+
+1. If none found, look into the `groups`  claim for user’s groups.
+
+> When attending to overage scenarios, which requires a call to [Microsoft Graph](https://graph.microsoft.com) to read the signed-in user's group memberships, you app will need to have the [Directory.Read.All](https://docs.microsoft.com/en-us/graph/permissions-reference#group-permissions) for the [getMemberObjects](https://docs.microsoft.com/en-us/graph/api/user-getmemberobjects?view=graph-rest-1.0) function to execute successfully.
+
+#### When using the implicit_grant flow to authenticate
+
+In case, you are authenticating using the [implicit grant flow](https://docs.microsoft.com/en-us/azure/active-directory/develop/v1-oauth2-implicit-grant-flow), the **overage** indication and limits are different than the apps using other flows.
+
+1. A claim named `hasgroups` with a value of true will be present in the token instead of the `groups` claim .
+1. The maximum number of groups provided in the `groups` claim is limited to 6. This is done to prevent  the URI fragment beyond the URL length limits.
+
+### Support in ASP.NET OWIN middleware libraries
+
+The asp.net middleware supports roles populated from claims by specifying the claim in the `RoleClaimType` property of `TokenValidationParameters`. 
+Since the `groups` claim contains the object ids of the security groups than actual names, the following code will not work. 
+
+```CSharp
+
+// Startup.Auth.cs
+public void ConfigureAuth(IAppBuilder app)
+{
+    app.SetDefaultSignInAsAuthenticationType(CookieAuthenticationDefaults.AuthenticationType);
+    app.UseCookieAuthentication(new CookieAuthenticationOptions());
+
+    //Configure OpenIDConnect, register callbacks for OpenIDConnect Notifications
+    app.UseOpenIdConnectAuthentication(
+    new OpenIdConnectAuthenticationOptions
+    {
+        ClientId = ConfigHelper.ClientId,
+        Authority = String.Format(CultureInfo.InvariantCulture, ConfigHelper.AadInstance, ConfigHelper.Tenant),
+                PostLogoutRedirectUri = ConfigHelper.PostLogoutRedirectUri,
+        TokenValidationParameters = new System.IdentityModel.Tokens.TokenValidationParameters
+        {
+            ValidateIssuer = false,
+            RoleClaimType = "groups",
```

Comment:
This code should be incrementally shown in the Startup.cs 

After:
```CSharp
            services.AddAzureAdV2Authentication(Configuration)
                    .AddMsal(new string[] { Constants.ScopeUserRead })
                    .AddInMemoryTokenCaches();
```

we can have

```CSharp
            services.Configure<OpenIdConnectOptions>(AzureADDefaults.OpenIdScheme, options =>
            {
             TokenValidationParameters.RoleClaimType = "groups";
            }
```

Reply:
This is turning out to be a bit more work than expected. Created #691139 for taking care of this later

---
In reply to: [272946796](https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2/pull/84#discussion_r272946796) [](ancestors = 272946796)